### PR TITLE
v4 external identifiers api test.

### PR DESCRIPTION
**Pull request by alxdarksage**

### CODE REVIEW

File: src/test/java/org/sagebionetworks/bridge/sdk/integration/ExternalIdsV3Test.java [hunk 1]
```diff
@@ -79,7 +79,7 @@ public void canCRUDExternalIds() throws IOException {
                     .execute().body();
             assertEquals(0, page1.getItems().size());
         } finally {
-            externalIdsClient.deleteExternalIds(identifiers).execute();
+            externalIdsClient.deleteExternalIdsV3(identifiers).execute();
         }
         ExternalIdentifierList page = externalIdsClient.getExternalIds(null, null, prefix, null).execute().body();
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/sdk/integration/ExternalIdsV3Test.java]: Should probably use getExternalIdsV3 to ensure coverage of the old APIs. Ideally, all calls in this class should be getExternalIdsV3.

File: src/test/java/org/sagebionetworks/bridge/sdk/integration/ExternalIdsV4Test.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+import org.sagebionetworks.bridge.rest.model.Substudy;
+import org.sagebionetworks.bridge.user.TestUserHelper;
+import org.sagebionetworks.bridge.user.TestUserHelper.TestUser;
+import org.sagebionetworks.bridge.util.IntegTestUtils;
+
+import com.google.common.collect.Lists;
+
+import retrofit2.Response;
+
+public class ExternalIdsV4Test {
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/sdk/integration/ExternalIdsV4Test.java]: Should we also have an assignment filter test here?

File: src/test/java/org/sagebionetworks/bridge/sdk/integration/ExternalIdsV4Test.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                researcherApi.createExternalId(new ExternalIdentifier().identifier(extIdA)).execute();
+                fail("Should have thrown an exception");
+            } catch (InvalidEntityException e) {
+            }
+
+            // Create a couple of external IDs related to different substudies.
+            ExternalIdentifier extId1 = new ExternalIdentifier().identifier(extIdA).substudyId(idA);
+            ExternalIdentifier extId2 = new ExternalIdentifier().identifier(extIdB).substudyId(idB);
+            researcherApi.createExternalId(extId1).execute();
+            researcherApi.createExternalId(extId2).execute();
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/sdk/integration/ExternalIdsV4Test.java]: We create 2 external IDs, but only use 1. Should we be doing something with the other one?

If this is just to test the creation of external IDs, should we get() the external IDs back to verify they were created?

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: src/test/java/org/sagebionetworks/bridge/sdk/integration/ExternalIdsV4Test.java (hunk 3 above)
Comment (alxdarksage) [reply src/test/java/org/sagebionetworks/bridge/sdk/integration/ExternalIdsV4Test.java]: Now that I have added testing of the filtering, the second one is used, so I will probably now leave it.

File: src/test/java/org/sagebionetworks/bridge/sdk/integration/ExternalIdsV4Test.java [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+            // Verify these were both created.
+            ExternalIdentifierList list = researcherApi.getExternalIds(null, null, null, false).execute().body();
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/sdk/integration/ExternalIdsV4Test.java]: We should use a prefix for this call, or else we might get random failures when the api study has other external IDs.

File: src/test/java/org/sagebionetworks/bridge/sdk/integration/ExternalIdsV4Test.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+            // Cannot create another user with this external ID. This should do nothing and fail quietly.
+            Response<Message> response = researcher.getClient(AuthenticationApi.class).signUp(signUp).execute();
+            assertEquals(201, response.code());
+
+            StudyParticipant participant2 = researcherApi.getParticipantByExternalId(extIdA, false).execute().body();
+            assertEquals(userId, participant2.getId());
+            
+            // verify filtering by assignment since we have assigned one record
+            ExternalIdentifierList all = researcherApi.getExternalIds(null, 50, null, null).execute().body();
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/sdk/integration/ExternalIdsV4Test.java]: We should use a prefix in this call (and in the next 2 calls), or we might get random failures when there are other external IDs in the api study.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

**Review by DwayneJengSage — APPROVED** [review]

**Pull request merged**